// event-dispatcher/gruntfile.js
module.exports = function (grunt) {
    "use strict";

    var defaultTasks;

    defaultTasks = ["jsbeautifier", "test"];

    grunt.initConfig({
        pkg: grunt.file.readJSON("package.json"),

        // grunt-contrib-clean
        clean: {
            instrument: "<%= instrument.options.basePath %>"
        },

        // grunt-contrib-jshint
        jshint: {
            files: [
                "<%= instrument.files %>",
                "<%= nodeunit.files %>",
                "gruntfile.js",
                "tests/**/*.js"
            ],
            options: grunt.file.readJSON(".jshintrc")
        },

        // grunt-contrib-nodeunit
        nodeunit: {
            files: ['tests/**/*Test.js']
        },

        // grunt-contrib-watch
        watch: {
            files: ["<%= jshint.files %>"],
            tasks: defaultTasks
        },

        // grunt-istanbul
        instrument: {
            files: [
                "module/*.js",
                "scripts/**/*.js"
            ],
            options: {
                basePath: "coverage/instrument/"
            }
        },
        storeCoverage: {
            options: {
                dir: "coverage/reports/<%= pkg.version %>"
            }
        },
        makeReport: {
            src: "<%= storeCoverage.options.dir %>/*.json",
            options: {
                type: "lcov",
                dir: "<%= storeCoverage.options.dir %>",
                print: "detail"
            }
        },

        // grunt-jsbeautifier
        jsbeautifier: {
            files: ["<%= jshint.files %>"],
            options: {
                js: grunt.file.readJSON(".jsbeautifyrc")
            }
        }
    });

    grunt.loadNpmTasks("grunt-contrib-clean");
    grunt.loadNpmTasks("grunt-contrib-jshint");
    grunt.loadNpmTasks("grunt-contrib-nodeunit");
    grunt.loadNpmTasks("grunt-contrib-watch");
    grunt.loadNpmTasks("grunt-istanbul");
    grunt.loadNpmTasks("grunt-jsbeautifier");

    grunt.registerTask("register_globals", function (task) {
        var moduleRoot;

        if ("coverage" === task) {
            moduleRoot = __dirname + "/" + grunt.template.process("<%= instrument.options.basePath %>");
        } else if ("test" === task) {
            moduleRoot = __dirname;
        }

        Object.defineProperty(global, "MODULE_ROOT", {
            enumerable: true,
            value: moduleRoot
        });
    });

    grunt.registerTask("beautify", ["jsbeautifier"]);
    grunt.registerTask("cover", ["register_globals:coverage", "instrument", "lint", "nodeunit", "storeCoverage", "makeReport", "clean:instrument"]);
    grunt.registerTask("lint", ["jshint"]);
    grunt.registerTask("test", ["register_globals:test", "clean:instrument", "lint", "nodeunit"]);

    grunt.registerTask("default", defaultTasks);
};
